File: apps/merchant-terminal-v2-android/app/src/main/java/com/yunqiao/life/merchantterminal/jobs/CanonicalServerPayload.kt
package com.yunqiao.life.merchantterminal.jobs

import com.yunqiao.life.merchantterminal.network.ClaimedV2PrintJob

object CanonicalServerPayload {
    private const val PROTOCOL = "ESC_POS_RASTER_V1"
    private const val TEMPLATE = "YQ_CANONICAL_RECEIPT_V1"

    fun forJob(job: ClaimedV2PrintJob, expectedWidthDots: Int): ByteArray? =
        when (job.renderProtocol) {
            null -> null
            PROTOCOL -> {
                require(job.canonicalTemplateVersion == TEMPLATE) {
                    "Canonical template version is unsupported."
                }
                require(job.widthDots == expectedWidthDots) {
                    "Canonical payload profile mismatch."
                }
                requireNotNull(job.renderedPayload) {
                    "Canonical payload is missing."
                }.copyOf()
            }
            else -> error("Unsupported server render protocol: ${job.renderProtocol}")
        }
}

File: clients/windows/YunQiao.Cashier/src/YunQiao.Cashier.Core/Protocol/CanonicalServerPayload.cs
namespace YunQiao.Cashier.Core.Protocol;

public static class CanonicalServerPayload
{
    private const string Protocol = "ESC_POS_RASTER_V1";
    private const string Template = "YQ_CANONICAL_RECEIPT_V1";

    public static byte[]? ForJob(ClaimedPrintJob job, int expectedWidthDots)
    {
        if (job.RenderProtocol is null) return null;
        if (job.RenderProtocol != Protocol)
            throw new InvalidDataException($"Unsupported server render protocol: {job.RenderProtocol}");
        if (job.CanonicalTemplateVersion != Template || job.WidthDots != expectedWidthDots)
            throw new InvalidDataException("Canonical payload profile mismatch.");
        return job.RenderedPayload?.ToArray()
            ?? throw new InvalidDataException("Canonical payload is missing.");
    }
}
